### PR TITLE
Unexpected behavior

### PR DESCRIPTION
**Pull request by riveSunder**
PR for a sub-branch off of branch `multihead` addressing issue #41. 

Summary: Fixed mis-specified softmax axis causing abberant transformer model output depending on batch size. 

Issue #41 was batch size dependent behavior during the forward pass: using a batch size (or single sequence) during inference time as input had high loss and nonsense output (_i.e._ an empty string or mono-character gibberish) despite low loss during training. 

Inference with the same batch size and with the batch functions used during training had good output. 

Systematic debugging led to the problem in `self_attention` function, softmax was specified for axis 0 which meant it was being applied w.r.t. each sample sequence. Fixed the softmax call in `self_attention` and in `encoded_attention` to specify axis -2, which is the sequence axis. Dimensions for batches are `bXsXt` where b is the batch size, s is the sequence length, and t is the token dimension. 

The fix is important and will be brought into master in  a separate PR. This PR is to keep branch `multihead` up to date with the fix while development of multi-head attention continues.

### CODE REVIEW

File: coverage.txt [hunk 1]
```diff
@@ -7,9 +7,9 @@ sepia/graph/data.py            13      7    46%
...
 sepia/seq/__init__.py           0      0   100%
-sepia/seq/data.py              78     36    54%
+sepia/seq/data.py              85     27    68%
 sepia/seq/dataloader.py        67     23    66%
 sepia/seq/functional.py       155      0   100%
-sepia/seq/transformer.py      207     54    74%
+sepia/seq/transformer.py      214     46    79%
 -----------------------------------------------
-TOTAL                         592    139    77%
+TOTAL                         606    122    80%
```
Comment (riveSunder) [inline coverage.txt:15]: ✓

File: sepia/seq/data.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 import numpy as np
 import numpy.random as npr
-aa_keys = "arndcqeghilkmfpstwyvuox".upper()
+aa_keys = "arndcqeghilkmfpstwyvuox"
```
Comment (riveSunder) [inline sepia/seq/data.py:7]: Previous logic implemented to parse strings as case-independent was incompatible with SMILES format strings (see [https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system](https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system))

File: sepia/seq/data.py [hunk 3]
```diff
@@ -117,6 +117,24 @@ def one_hot_to_sequence(one_hot: jnp.array, sequence_dict: dict) -> str:
 
     return sequence
 
+def batch_one_hot_to_sequence(one_hot: jnp.array, sequence_dict: dict) -> list:
```
Comment (riveSunder) [inline sepia/seq/data.py:120]: As part of this branch I tried to minimize any difference in the forward pass used by string inference by `__call__` and the forward pass used during training. This should make debugging easier. 

`__call__` _i.e._ `model(sequence)` now takes string sequences or lists of strings. Lists of string sequences are preferred (note: update the type hint in `__call__`)

File: sepia/seq/data.py [hunk 4]
```diff
@@ -117,6 +117,24 @@ def one_hot_to_sequence(one_hot: jnp.array, sequence_dict: dict) -> str:
...
+    """
+    process a batch of sequences represented by one hot encodings
+    returns a list of strings
+    """
+
+    shape_length = len(one_hot.shape)
+    assert shape_length == 3, f"expected 3 dims in one_hot batch vector, got {shape_length}"
+
+    output = []
+    #loop through the
```
Comment (riveSunder) [inline sepia/seq/data.py:130]: Please finish or delete this comment

File: sepia/seq/functional.py [hunk 5]
```diff
@@ -74,7 +74,7 @@ def self_attention(x: jnp.array, parameters: SelfAttentionW) -> jnp.array:
     dim_k = query.shape[-1]
     raw_attention = jnp.matmul(query, key.transpose(0,2,1)) / jnp.sqrt(dim_k)
 
-    attention = jax.nn.softmax(raw_attention, axis=0)
+    attention = jax.nn.softmax(raw_attention, axis=-2)
```
Comment (riveSunder) [inline sepia/seq/functional.py:77]: This (and one other, identical, line used elsewhere) is the critical fix for this PR.

File: sepia/seq/functional.py [hunk 6]
```diff
@@ -122,7 +122,7 @@ def encoded_attention(x: jnp.array, encoded: jnp.array, parameters: EncodedAtten
 
     raw_attention = jnp.matmul(encoded_query, encoded_key.transpose(0,2,1)) / jnp.sqrt(dim_k)
 
-    attention = jax.nn.softmax(raw_attention, axis=0)
+    attention = jax.nn.softmax(raw_attention, axis=-2)
```
Comment (riveSunder) [inline sepia/seq/functional.py:125]: This (and one other, identical, line used elsewhere) is the critical fix for this PR.

**Review by riveSunder — COMMENTED** [review]
Please implement a few changes before merging this PR:
in `sepia/seq/data.py`:
* finish or delete comment at line 130
in `sepia/seq/transformer.py`:
* update type hint in `__call__`. A list of strings is the preferred input format

File: sepia/seq/data.py [hunk 7]
```diff
@@ -117,6 +117,24 @@ def one_hot_to_sequence(one_hot: jnp.array, sequence_dict: dict) -> str:
...
+    """
+    process a batch of sequences represented by one hot encodings
+    returns a list of strings
+    """
+
+    shape_length = len(one_hot.shape)
+    assert shape_length == 3, f"expected 3 dims in one_hot batch vector, got {shape_length}"
+
+    output = []
+    # loop through each sequence in batch
```
Comment (riveSunder) [inline sepia/seq/data.py:130]: ✓

File: sepia/seq/transformer.py [hunk 8]
```diff
@@ -311,59 +314,49 @@ def encode(self, sequence: str) -> jnp.array:
 
         return encoded
 
-
-    def __call__(self, sequence: str) -> str:
+    def __call__(self, sequence: list) -> str:
```
Comment (riveSunder) [inline sepia/seq/transformer.py:317]: ✓

**Review by riveSunder — COMMENTED** [review]
OK ✓
